newIDE/app/src/EventsFunctionsList/index.js

```diff
@@ -6,509 +6,1211 @@ import { t } from '@lingui/macro';
+                        ),
+                      ]
+                    : objectTreeViewItems;
+                },
+              }
+            : null,
+          {
+            isRoot: true,
+            content: new LabelTreeViewItemContent(
+              extensionBehaviorsRootFolderId,
+              i18n._(t`Behaviors`),
+              {
+                icon: <Add />,
+                label: i18n._(t`Add a behavior`),
+                click: addNewEventsBehavior,
+              }
+            ),
+            getChildren(i18n: I18nType): ?Array<TreeViewItem> {
+              return behaviorTreeViewItems.length === 0
+                ? [
+                    new PlaceHolderTreeViewItem(
+                      extensionBehaviorsEmptyPlaceholderId,
+                      i18n._(t`Start by adding a new behavior.`)
+                    ),
+                  ]
+                : behaviorTreeViewItems;
+            },
+          },
+          {
+            isRoot: true,
+            content: new LabelTreeViewItemContent(
+              extensionFunctionsRootFolderId,
+              i18n._(t`Functions`),
+              {
+                icon: <Add />,
+                label: i18n._(t`Add a function`),
+                click: () => addNewEventsFunction(null),
+              }
+            ),
+            getChildren(i18n: I18nType): ?Array<TreeViewItem> {
+              if (eventsFunctionsExtension.getEventsFunctionsCount() === 0) {
+                return [
+                  new PlaceHolderTreeViewItem(
+                    extensionFunctionsEmptyPlaceholderId,
+                    i18n._(t`Start by adding a new function.`)
+                  ),
+                ];
+              }
+              const freeFunctionProps = {
+                eventsFunctionsContainer: eventsFunctionsExtension,
+                ...eventFunctionCommonProps,
+              };
+              return mapFor(
+                0,
+                eventsFunctionsExtension.getEventsFunctionsCount(),
+                i =>
+                  new LeafTreeViewItem(
+                    new FunctionTreeViewItemContent(
+                      eventsFunctionsExtension.getEventsFunctionAt(i),
+                      freeFunctionProps
+                    )
+                  )
+              );
+            },
+          },
+        ].filter(Boolean);
+      },
+      [
+        addNewEventsBehavior,
+        addNewEventsFunction,
+        addNewEventsObject,
+        behaviorTreeViewItems,
+        eventFunctionCommonProps,
+        eventsFunctionsExtension,
+        objectTreeViewItems,
+        getShowEventBasedObjectsEditor,
+      ]
+    );
 
-  render() {
-    const {
-      project,
-      eventsFunctionsContainer,
-      selectedEventsFunction,
-      onSelectEventsFunction,
-    } = this.props;
-    const { searchText } = this.state;
+    const canMoveSelectionTo = React.useCallback(
+      (destinationItem: TreeViewItem, where: 'before' | 'inside' | 'after') =>
```

Comment:
I didn't used the `TreeViewItemContent` for this function but it may be easier to read as it would split the logic by targeted item type. I guess it's small enough to avoid to split it.

Reply:
Polymorphism would tell us to move this into a method, but yeah it's ok as it's fairly small I guess! :)